_compdemos/Cromwell.md

```diff
@@ -4,18 +4,101 @@ primary_reviewers: vortexing
+
+#### Docker and Singularity
+For all of those, however, who do not have access to our SLURM cluster and all those magic software modules, it means that for them, our workflows still aren't directly portable or shareable.  For that scenario, a containerization approach like using [Docker](/scicomputing/compute_environments/#docker-containers).  What this looks like in the context of workflows is that every task described in a workflow is given as a runtime variable a Docker container name and tag which specifies the exact container to run the task inside that will have all the software required for the task to be performed.  In the case of workflows running on our local SLURM cluster, that means that Cromwell is configured to use Singularity in order to pull and convert Docker images prior to running the task.  Meanwhile, on cloud based backends like AWS Batch, it means that the Docker container can be pulled and used without conversion.  Designing workflows to use publicly available Docker containers provides a substantial improvement in the likelihood that another user or collaborator would be able to run your workflow, reproducibly, on our or their own HPC resources.  
+
+
+### Bells and Whistles of Workflow Managers
+Beyond the issues associated with reproducibility and the burden of institution-specific computing knowledge, using a workflow manager like Cromwell has a number of other advantages that can make your work much faster, more efficient and scalable. 
+
+#### Call Caching and Workflow Restarts
+Call caching is a major strength of workflow managers and essentially is a function performed by the workflow manager software that identifies, for a given task, what the inputs, task instructions and expected outputs are, and queries it's database to determine if that exact combination has been successfully done previously.  In the event that an exact match is found, then the workflow manager finds the output of the previously run task and immediately moves forward in the workflow, avoiding re-computing that output.  
+
+This is a valuable benefit because often when a complex or highly parallel workflow is run, there may be a variety of individual task failures due to problems that aren't in the code itself but rather things like internet connectivity problems or other intermittent problems.  Workflow managers can manage both the direct resubmission of individual failed tasks that failed due to HPC resource failures, as well as allow the user to restart an entire workflow should a task fail.  This means that work can be done more efficiently even when individual tasks might fail, because when the user might provide a fix or even directly resubmit a workflow, only those tasks impacted will be re-run, not those that completed successfully.  
```

Comment:
This means that work can be done more efficiently even if individual tasks fail, because when the user might provide a fix or even directly resubmit a workflow, only those tasks impacted will be re-run, not those that completed successfully.